src/chips/sam4l/spi.rs

```diff
@@ -0,0 +1,284 @@
+        match self.get_active_peripheral() {
+            Peripheral::Peripheral0 => volatile!(self.regs.csr0 = value),
+            Peripheral::Peripheral1 => volatile!(self.regs.csr1 = value),
+            Peripheral::Peripheral2 => volatile!(self.regs.csr2 = value),
+            Peripheral::Peripheral3 => volatile!(self.regs.csr3 = value),
+        };
+    }
+
+
+    /// Ends the SPI transaction by setting the slave select high
+    fn end_transaction(&mut self) {
+        volatile!(self.regs.cr = 1 << 24);
+    }
+}
+
+impl spi_master::SPI for SPI {
+    fn init(&mut self, params: spi_master::SPIParams) {
+        self.client = params.client;
+        self.set_baud_rate(params.baud_rate);
+
+        let mut csr = self.read_active_csr();
+        // Clock polarity
+        match params.clock_polarity {
+            spi_master::ClockPolarity::IdleHigh => csr |= 1, // Set bit 0
+            spi_master::ClockPolarity::IdleLow => csr &= 0xFFFFFFFE, // Clear bit 0
+        };
+        // Clock phase
+        match params.clock_phase {
+            spi_master::ClockPhase::SampleLeading => csr |= 1 << 1, // Set bit 1
+            spi_master::ClockPhase::SampleTrailing => csr &= 0xFFFFFFFD, // Clear bit 1
+        }
+        // Keep slave select active until a last transfer bit is set
+        csr |= 1 << 3;
+        self.write_active_csr(csr);
+
+        // Indicate the last transfer, so that the slave select will be disabled
+        volatile!(self.regs.cr = 1 << 24);
+
+        let mut mode = volatile!(self.regs.mr);
+        // Enable master mode
+        mode |= 1;
+        // Disable mode fault detection (open drain outputs do not seem to be supported)
+        mode |= 1 << 4;
+        volatile!(self.regs.mr = mode);
+    }
+
+    fn write_byte(&mut self, out_byte: u8, last_transfer: bool) -> u8 {
+        let tdr = out_byte as u32;
+        volatile!(self.regs.tdr = tdr);
+        if last_transfer {
+            self.end_transaction();
+        }
+        // Wait for receive data register full
+        while (volatile!(self.regs.sr) & 1) != 1 {}
+        // Return read value
+        volatile!(self.regs.rdr) as u8
+    }
+
+    fn read_byte(&mut self, last_transfer: bool) -> u8 {
+        self.write_byte(0, last_transfer)
+    }
+
+    fn read(&mut self, buffer: &mut [u8], last_transfer: bool) {
+        // TODO: Asynchronous
+        for i in 0..buffer.len() {
+            // Write 0
+            let tdr: u32 = 0;
+            volatile!(self.regs.tdr = tdr);
+            if last_transfer && i == buffer.len() - 1 {
+                self.end_transaction();
+            }
+            // Wait for receive data register full
+            while (volatile!(self.regs.sr) & 1) != 1 {}
+
+            buffer[i] = volatile!(self.regs.rdr) as u8;
+        }
+        if let Some(ref mut client) = self.client {
+            client.read_done();
+        }
+    }
+
+    fn write(&mut self, buffer: &[u8], last_transfer: bool) {
```

Comment:
I strongly suspect that when we actually try to make this asynchronous, the current type signature will not allow us to borrow `buffer` for long enough. This goes for the other functions that should eventually made asynchronous.

I think there are two potential ways around this:

1. Try to use `AppPtr` (or more likely `AppSlice`) here. Those have unbounded lifetimes (since they are basically dynamically allocated) so don't have a similar problem. High-level device drivers, which have the ability to create such pointers, are most likely going to be the clients of this interface, so that could work out just fine.

2. Configure &'static [u8] slices during initialization.